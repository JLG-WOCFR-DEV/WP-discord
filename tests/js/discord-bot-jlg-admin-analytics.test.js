beforeAll(() => {
    global.__DISCORD_BOT_JLG_DISABLE_AUTO_BOOTSTRAP = true;
});

const analytics = require('../../discord-bot-jlg/assets/js/discord-bot-jlg-admin-analytics.js');

describe('discord-bot-jlg-admin-analytics helpers', () => {
    test('prepareConfig fills defaults and keeps snapshots stable', () => {
        const config = analytics.prepareConfig({
            profiles: [
                { key: 'alpha', label: 'Alpha' },
                { key: 'beta' }
            ],
            defaultProfiles: ['alpha'],
            requestedProfiles: ['alpha', 'beta'],
            comparisonPresets: [
                { id: 'solo-alpha', label: 'Alpha only', profiles: ['alpha'] }
            ],
            annotations: [
                { timestamp: 1700000000, label: 'Launch', profile_key: 'alpha', metric: 'presence' }
            ]
        });

        expect(config).toMatchSnapshot();
    });

    test('buildDatasets generates aligned datasets with annotations', () => {
        const timeline = [100, 200, 300];
        const seriesMap = {
            alpha: {
                label: 'Alpha',
                timeseries: [
                    { timestamp: 100, presence: 10, online: 5, total: 50, premium: 1 },
                    { timestamp: 200, presence: 15, online: 6, total: 55, premium: 1 },
                    { timestamp: 300, presence: 12, online: 7, total: 60, premium: 2 }
                ]
            },
            beta: {
                label: 'Beta',
                timeseries: [
                    { timestamp: 100, presence: 20, online: 8, total: 70, premium: 0 },
                    { timestamp: 300, presence: 24, online: 9, total: 73, premium: 1 }
                ]
            }
        };

        const annotations = [
            { timestamp: 200, label: 'Milestone', profiles: ['alpha'], metric: 'presence' },
            { timestamp: 300, label: 'Campaign', profiles: ['*'], metric: '*' }
        ];

        const result = analytics.buildDatasets({
            timeline,
            seriesMap,
            activeProfiles: ['alpha', 'beta'],
            metric: 'presence',
            showAnnotations: true,
            labels: {
                metricPresence: 'PrÃ©sence',
                annotations: 'Annotations'
            },
            annotations
        });

        expect(result.datasets.length).toBe(2);
        expect(result).toMatchSnapshot();
    });

    test('buildExportRows flattens multi-profile data', () => {
        const seriesMap = {
            alpha: {
                timeseries: [
                    { timestamp: 100, online: 5, presence: 10, total: 50, premium: 1 },
                    { timestamp: 200, online: 6, presence: 12, total: 52, premium: 1 }
                ]
            },
            beta: {
                timeseries: [
                    { timestamp: 100, online: 8, presence: 20, total: 70, premium: 0 }
                ]
            }
        };

        const rows = analytics.buildExportRows(seriesMap, [100, 200], ['alpha', 'beta']);
        expect(rows).toEqual([
            ['profile_key', 'timestamp', 'datetime', 'online', 'presence', 'total', 'premium'],
            ['alpha', 100, new Date(100 * 1000).toISOString(), 5, 10, 50, 1],
            ['beta', 100, new Date(100 * 1000).toISOString(), 8, 20, 70, 0],
            ['alpha', 200, new Date(200 * 1000).toISOString(), 6, 12, 52, 1]
        ]);
    });
});
