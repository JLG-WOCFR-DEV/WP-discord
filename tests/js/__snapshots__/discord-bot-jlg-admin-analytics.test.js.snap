// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`discord-bot-jlg-admin-analytics helpers buildDatasets generates aligned datasets with annotations 1`] = `
{
  "annotationMeta": {
    "0": {
      "1": [
        "Annotations • 15 • 1/1/1970, 12:03:20 AM",
        "Milestone • 1/1/1970, 12:03:20 AM",
      ],
      "2": [
        "Campaign • 1/1/1970, 12:05:00 AM",
      ],
    },
    "1": {
      "2": [
        "Annotations • 24 • 1/1/1970, 12:05:00 AM",
        "Campaign • 1/1/1970, 12:05:00 AM",
      ],
    },
  },
  "datasets": [
    {
      "backgroundColor": "rgba(29, 78, 216, 0.16)",
      "borderColor": "#1d4ed8",
      "data": [
        10,
        15,
        12,
      ],
      "fill": false,
      "label": "Alpha — Présence",
      "pointHoverRadius": [
        5,
        9,
        9,
      ],
      "pointRadius": [
        2,
        7,
        7,
      ],
      "spanGaps": true,
      "tension": 0.3,
    },
    {
      "backgroundColor": "rgba(22, 163, 74, 0.16)",
      "borderColor": "#16a34a",
      "data": [
        20,
        null,
        24,
      ],
      "fill": false,
      "label": "Beta — Présence",
      "pointHoverRadius": [
        5,
        5,
        9,
      ],
      "pointRadius": [
        2,
        2,
        7,
      ],
      "spanGaps": true,
      "tension": 0.3,
    },
  ],
}
`;

exports[`discord-bot-jlg-admin-analytics helpers prepareConfig fills defaults and keeps snapshots stable 1`] = `
{
  "annotations": [
    {
      "label": "Launch",
      "metric": "presence",
      "profiles": [
        "alpha",
      ],
      "timestamp": 1700000000,
    },
  ],
  "canvasId": "discord-analytics-chart",
  "comparisonPresets": [
    {
      "id": "solo-alpha",
      "label": "Alpha only",
      "profiles": [
        "alpha",
      ],
    },
  ],
  "containerId": "discord-analytics-panel",
  "days": 7,
  "defaultMetric": "presence",
  "defaultProfiles": [
    "alpha",
  ],
  "labels": {},
  "nonce": "",
  "profiles": [
    {
      "key": "alpha",
      "label": "Alpha",
      "server_id": "",
    },
    {
      "key": "beta",
      "label": "beta",
      "server_id": "",
    },
  ],
  "requestedProfiles": [
    "alpha",
    "beta",
  ],
  "restUrl": "",
}
`;
